### DISCUSSION TITLE
SRV Component Mapping

### DISCUSSION THREAD

**Discussion by RobertMtx** [DiligentGraphics/DiligentEngine#228] [Q&A] [answered]
Is there any way to use component mapping while using Diligent? I tried to look into it, but from what I can tell, it looks like all of the values are being set to default for DirectX:

`d3d12SRVDesc.Shader4ComponentMapping = D3D12_DEFAULT_SHADER_4_COMPONENT_MAPPING;`

Would it be feasible to modify this value without a ton of effort? It would be super useful to set this up as `RRR1` for greyscale textures, so that they can be sampled along with diffuse RGB textures, without the shader code needing to know what format it is dealing with. Currently, I am having to compile separate versions of every shader that supports compressed greyscale textures in a certain slot.

It looks like both DirectX and OpenGL support this feature. From my limited research, it looks like the OpenGL version is called "Swizzle mask".

**Comment by TheMostDiligent** [comment, marked as answer]
Yes, there is an [open issue for this feature](https://github.com/DiligentGraphics/DiligentCore/issues/258)

**Reply by TheMostDiligent** [reply]
FYI, I added texture component mapping to [the API](https://github.com/DiligentGraphics/DiligentCore/blob/master/Graphics/GraphicsEngine/interface/TextureView.h#L218)

**Comment by RobertMtx** [comment]
That's amazing, thanks!

**Comment by RobertMtx** [comment]
Is there any way to apply this component mapping to the default shader view of a texture that is created through the loader? Or is it necessary to create a secondary view of any texture we want to swizzle the components of?

Or another question might be - is it possible to customize the default shader view of textures? The swizzle component being in the texture view description makes it appear difficult to reach without creating additional views, but maybe I'm missing something?

Lastly, if swizzle does require a custom view, would adding additional views to *all* textures have any significant cost?

Thanks again for this feature. It will be a life saver when using compressed block formats and grayscale images.

**Reply by TheMostDiligent** [reply]
> Is there any way to apply this component mapping to the default shader view of a texture that is created through the loader

Which loader do you mean? Default view of a texture addresses the entire texture, which is why it is default.

> is it possible to customize the default shader view of textures

To customize the default view you can create your own view.

> Lastly, if swizzle does require a custom view, would adding additional views to all textures have any significant cost?

Texture view is a handle. It does require some memory on the order of few tens of bytes. Compared to the size of the texture this is usually negligible.

**Reply by RobertMtx** [reply]
> > Is there any way to apply this component mapping to the default shader view of a texture that is created through the loader
> 
> Which loader do you mean? Default view of a texture addresses the entire texture, which is why it is default.

I was referring to the texture loader (`Diligent::ITextureLoader`). This is what I've always used to create textures through Diligent, but it doesn't appear to allow any inputs to control the component mapping or to influence the resulting `TextureViewDesc`, apart from the  inputs of `Diligent::TextureLoadInfo`. But it sounds like creating another view is the way to go. Thanks.

**Reply by RobertMtx** [reply]
I just tested the component swizzle feature out, and it works great. Thanks again for adding that in.

My game will make pretty extensive use of this. For example, my font atlas textures are grayscale BC4 (edited) compressed textures, and swizzle to 1,1,1,R to use their value as an alpha mask. Without this feature, I had to use a custom shader for font rendering, which required switching shaders every time I drew text in between 2D images. But with this, I can render large chunks of 2D images and fonts, all together in a single draw call. My frame rate went from 180 to 240 with only about a quarter of the screen filled with UI elements.

EDIT: I stated BC5, but meant BC4

**Reply by TheMostDiligent** [reply]
> I was referring to the texture loader (Diligent::ITextureLoader)

Texture loader only creates the texture. Swizzle is part of the texture view, so yes, you need to create an appropriate view after you load the texture from file.
